### PR TITLE
beginnings of fluent arcs out & back from ValidationRules class

### PR DESCRIPTION
**Pull request by floyd-may**
Hopefully this is a bit of a start in the right direction.

### CODE REVIEW

File: Assisticant.UnitTest/NotifyDataErrorInfoTests.cs [hunk 1]
```diff
@@ -56,14 +56,16 @@ public DateTime Birth
                 set { _death.Value = value; }
             }
 
-            public ValidationRules Rules => new ValidationRules()
-                .ForString(() => PhoneNumber, r => r
-                    .Matches(@"^[0-9\-\(\)]*$"))
+            public IValidationRules Rules => new ValidationRules()
+                .For(() => PhoneNumber)
+                    .Matches(@"^[0-9\-\(\)]*$")
```
Comment (floyd-may) [inline Assisticant.UnitTest/NotifyDataErrorInfoTests.cs]: This `For(...)` arc is strongly-typed to string, so the `Matches(...)` method is only available for strings.

**Comment by floyd-may** [comment]
This should be a reasonable solution for the discussion in #22.

**Comment by michaellperry** [comment]
This syntax requires that I re-express a property in order to supply multiple rules. For example, see the branch pr23-reexpress. (Sorry, I don't know how to comment on a PR with another change.)

I'd like to use the language semantics to represent the hierarchy, and not rely upon the user to indent expressions that are syntactically on the same level.

**Comment by floyd-may** [comment]
> This syntax requires that I re-express a property in order to supply multiple rules.

Let me chew on that for a bit. I've got a couple ideas but it might require decoupling the `ValidationRules` class from the fluent stuff. Are you opposed to having a `.Build()` at the end of the chain?

**Comment by michaellperry** [comment]
Not opposed at all.

**Comment by floyd-may** [comment]
> I'd like to use the language semantics to represent the hierarchy, and not rely upon the user to indent expressions that are syntactically on the same level.

Understood. Indentation isn't a "must" but it can be helpful. I'm thinking along the same lines as [what Fluent Migrator does](https://github.com/schambers/fluentmigrator/wiki/Fluent-Interface). When I write Fluent Migrator code, I indent the subsequent fluent calls corresponding to each column because I hate long lines.

**Comment by floyd-may** [comment]
Okay, @michaellperry, how does this version grab you? You can chain on as many different validation rules as you like, each with their own message, to a single property.

**Comment by michaellperry** [comment]
Very nice. Your use of inheritance with PropertyValidationContext lets the graph return home at any time, yet remain type safe while configuring a property.

Let me build the NuGet package and kick the tires a bit. If it all looks good in practice, there are a couple of features I'd like to add:

- Disable all validations until the user tries to save the form
- Disable validations for sets of properties based on other selections

I think I can get both of these with the same syntax and some clever use of your PropertyRuleset collections.

**Comment by michaellperry** [comment]
Noticed a couple of things:

- A standard validation rule cannot be followed by .WithMessage in case you wanted to override the standard message with a localized string.
- You cannot follow a generic .Where validation rule with a standard rule, because it looses its specific type.

I think the first is easily fixable. The second one might be something we just have to live with. I'll try a few things.

**Comment by michaellperry** [comment]
I opened a PR on your PR with what I found. Replacing the derived classes with extension methods gives us both benefits. I was frankly surprised that it worked!

Now it's just a simple matter of making the .WithMessage optional and providing a default message.

**Comment by floyd-may** [comment]
> I opened a PR on your PR
LOL that's awesome!
image

**Pull request merged**